projects/frontend/src/draft-posts/index.md

```diff
@@ -0,0 +1,148 @@
+		director(req)
+		req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
+		req.Host = req.URL.Host
+	}
+
+	http.HandleFunc("/", proxy.ServeHTTP)
+	log.Printf("Listening on port %d", *port)
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
+}
+```
+
+That’s it! This server can now handle proxying HTTP 1 and 2 as well as web socket connections. You’ll notice that I’ve configured the _proxy.Director_ field. The _ReverseProxy.Director_ is a function that modifies the incoming request before it is forwarded. The signature is as follows:
+
+```go
+Director func(*http.Request)
+```
+
+A common use case for the director function is altering request headers. There is a default director set by  _httputil.NewSingleHostReverseProxy_ . The default implementation takes care of setting the request scheme, host, and path. I found this useful and didn’t want to duplicate the code. I had to reset the _req.Host_  field to contact HTTPS endpoints . I’ve also included an example of setting a request header.
+
+## Capturing Metrics 
+Not too long ago I was asked to build a demo for one of our clients at work. The task was to capture the response metrics of a web service and add those metrics to [Prometheus](https://prometheus.io/) for display in [Grafana](https://grafana.com/). The only catch was that I did not have direct access to the web service and therefore couldn’t instrument the code. I completed the task by adapting the go reverse proxy to observe the network traffic coming from the service.
+
+Let’s extend our simple proxy to read and report metrics about the downstream service responses. To do this we’ll return to the  _httputil.ReverseProxy_ struct once more. It exposes a  struct field _ReverseProxy.ModifyResponse_ which gives us access to the http response before it goes back to the client.
+
+```go
+ModifyResponse func(*net/http.Response) error
+```
+
+HTTP bodies in go are implemented as _io.Reader_’s and therefore they can only be read once. If you would like to parse them before forwarding the request/response you will need to copy the body into a byte buffer and the reset the request/response body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.
+
+```go
+func duplicateReader(in io.ReadCloser) (b, b1 io.ReadCloser, err error) {
+	bBytes, err := ioutil.ReadAll(in)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	in.Close()
+
+	buf := bytes.NewBuffer(bBytes)
+	b = ioutil.NopCloser(buf)
+	b1 = ioutil.NopCloser(bytes.NewBuffer(bBytes))
+
+	return
+}
+```
+
+Depending on your use case you may not actually want to duplicate the reader. You may opt for returning the parsed body of the response i.e. parsing the json into a struct or _map[string]interface{}_  and just resetting _req.Body_ to an equivalent reader.
```

Comment:
I would potentially go through the trouble of moving this to the end and giving that implementation. Or leaving it commented in the code snippet